src/Features/Core/Portable/SplitOrMergeIfStatements/Consecutive/AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider
+        : AbstractMergeIfStatementsCodeRefactoringProvider
+    {
+        // Converts:
+        //    if (a)
+        //        Console.WriteLine();
+        //    else if (b)
+        //        Console.WriteLine();
+        //
+        // To:
+        //    if (a || b)
+        //        Console.WriteLine();
+
+        // Converts:
+        //    if (a)
+        //        return;
+        //    if (b)
+        //        return;
+        //
+        // To:
+        //    if (a || b)
+        //        return;
+
+        protected sealed override CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument, string ifKeywordText)
+            => new MyCodeAction(createChangedDocument, ifKeywordText);
+
+        protected sealed override async Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifLikeStatement, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+
+            return CanBeMergedWithParent(syntaxFacts, ifGenerator, ifLikeStatement) ||
+                   await CanBeMergedWithPreviousStatementAsync(document, syntaxFacts, ifGenerator, ifLikeStatement, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected sealed override SyntaxNode GetChangedRoot(Document document, SyntaxNode root, SyntaxNode ifLikeStatement)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+            var generator = document.GetLanguageService<SyntaxGenerator>();
```

Comment:
i suggest a helper that takes a document and returns a tuple :)

Reply:
I don't like the idea of introducing a new base class just for this one helper.